src/descriptor/mod.rs

```diff
@@ -509,6 +510,64 @@ impl<Pk: MiniscriptKey + ToPublicKey> Descriptor<Pk> {
     }
 }
 
+impl Descriptor<DefiniteDescriptorKey> {
+    /// Returns a plan if the provided assets are sufficient to produce a non-malleable satisfaction
+    ///
+    /// If the assets aren't sufficient for generating a Plan, the descriptor is returned
+    pub fn plan<P>(self, provider: &P) -> Result<Plan, Self>
+    where
+        P: AssetProvider<DefiniteDescriptorKey>,
+    {
+        let satisfaction = match self {
+            Descriptor::Bare(ref bare) => bare.plan_satisfaction(provider),
+            Descriptor::Pkh(ref pkh) => pkh.plan_satisfaction(provider),
+            Descriptor::Wpkh(ref wpkh) => wpkh.plan_satisfaction(provider),
+            Descriptor::Wsh(ref wsh) => wsh.plan_satisfaction(provider),
+            Descriptor::Sh(ref sh) => sh.plan_satisfaction(provider),
+            Descriptor::Tr(ref tr) => tr.plan_satisfaction(provider),
+        };
```

Comment:
In d29c2984e60e41e3872ac23b8b59e8b25fea88a2:

In a followup PR, these are easy candidates to un-recursify (using iter.rs from #567 which was created long after this PR but merged long before).

Reply:
For more context, what he means is what I'm doing in https://github.com/rust-bitcoin/rust-miniscript/pull/594. The `iter` module implements a pretty hard core data structure, if you don't have the time or inclination to work it out right now just holla at me and I'll hack it up for you.